### ISSUE TITLE
JSON parsing error

### ISSUE THREAD

**Issue by abhishek-shori** [couchbase/couchbase-lite-ios#1347]
Hi, 

I've run into a peculiar problem with my continuous pull replication. The replication is always successful for the first time when the app is launched. It is only after it finishes, it stops working when I expect changes in the app. 

I get the following error message. 
```
13:23:59.744‖ ChangeTracker: CBLSocketChangeTracker[0x7ff80b44f080 clientstore_2_ae5b4c6f-fd0a-4534-bc05-6a1a5d72c3f9]: Starting...
13:23:59.744‖ SyncVerbose: CBLSocketChangeTracker[0x7ff80b44f080 clientstore_2_ae5b4c6f-fd0a-4534-bc05-6a1a5d72c3f9]: POST //192.168.99.100:7784/clientstore_2_ae5b4c6f-fd0a-4534-bc05-6a1a5d72c3f9/_changes
13:23:59.744‖ ChangeTracker: Changes feed using proxy settings {
    ExceptionsList =     (
        "*.local",
        "169.254/16"
    );
    FTPPassive = 1;
    "__SCOPED__" =     {
        en0 =         {
            ExceptionsList =             (
                "*.local",
                "169.254/16"
            );
            FTPPassive = 1;
        };
    };
}
13:23:59.745‖ ChangeTracker: CBLSocketChangeTracker[0x7ff80b44f080 clientstore_2_ae5b4c6f-fd0a-4534-bc05-6a1a5d72c3f9]: Started... <http://192.168.99.100:7784/clientstore_2_ae5b4c6f-fd0a-4534-bc05-6a1a5d72c3f9/_changes>
13:23:59.745‖ ChangeTracker: CBLSocketChangeTracker[0x7ff80b44f080 clientstore_2_ae5b4c6f-fd0a-4534-bc05-6a1a5d72c3f9]: Event 1 on <__NSCFInputStream: 0x7ff808cd7400>
13:23:59.749‖ 

**WARNING: JSON error parsing _changes feed: lexical error: invalid char in son text.**
or this 
**WARNING: JSON error parsing _changes feed: parse error: after key and value, inside map, I expect ',' or '}'**
or 
**WARNING: JSON error parsing _changes feed: parse error: after array element, I expect ',' or ']'**

13:23:59.749‖ CBLSocketChangeTracker[0x7ff80b44f080 clientstore_2_ae5b4c6f-fd0a-4534-bc05-6a1a5d72c3f9]: Connection error #1, retrying in 2.0 sec: The operation couldn’t be completed. (CBLChangeTracker error 587.)
```
Another type of error:
```
Changes feed using proxy settings {
    FTPPassive = 1;
    HTTPEnable = 1;
    HTTPPort = 55848;
    HTTPProxy = "127.0.0.1";
    HTTPSEnable = 1;
    HTTPSPort = 55848;
    HTTPSProxy = "127.0.0.1";
    "__SCOPED__" =     {
        en0 =         {
            FTPPassive = 1;
            HTTPEnable = 1;
            HTTPPort = 55848;
            HTTPProxy = "127.0.0.1";
            HTTPSEnable = 1;
            HTTPSPort = 55848;
            HTTPSProxy = "127.0.0.1";
        };
    };
}
```

Thanks!
--------
* Version:    1.2.1
* Client OS: iOS 8.0
* Xcode :     7.3
* Server:      CouchDB 1.6.1

**Comment by snej** [comment]
For some reason it's receiving invalid JSON in the response from `/db/_changes`. I've seen this happen when a proxy server or gateway closes the socket, but then it's an unexpected end of data, while the errors reported here are about invalid characters, which is weird.

This sounds like a problem with the connection between CBL and CouchDB; I'm guessing something is writing unexpected data into the HTTP response for some reason. Do you have any proxies/gateways, like nginx or an AWS load balancer, in front of the CouchDB server?

**Comment by snej** [comment]
Oh, and the dump of the proxy settings in the log is showing that you have a proxy running on localhost — is it Charles Proxy or some similar debugging tool? Are you sure it isn't modifying the traffic?

**Comment by abhishek-shori** [comment]
Hey! Thanks for the prompt reply. I don't have any proxies/gateways running, & I only ran Charles to look at the /db/_changes responses. 
I took the unreleased 1.3.0, & somehow it worked. But I still want to make it work with 1.2.1 So please don't close this yet & give me a day more. Let me come back to you. Thanks!

**Comment by abhishek-shori** [comment]
And kudos to the team! Replication with 1.3 is super fast. Replicated some 100K docs in mins.

**Comment by snej** [comment]
Oh, cool. We've put in some optimizations on both client and server; glad they're noticeable :)

If you can read the JSON response, do you see the part that's invalid? If so, what does it look like?

**Comment by abhishek-shori** [comment]
} is missing the end. That's it.

**Comment by snej** [comment]
Huh. It sounds like the socket got closed before the rest of the response was sent. Do you know if the request had been active for a long time? Were there any actual JSON change objects in the response?

**Comment by abhishek-shori** [comment]
hey!

Thanks to you, I believe I've figured this out. I am running CouchDb inside Docker, Now the docker daemon on macOS can't run natively, instead it is hosted by a Linux VM. So it might be breaking the interaction b/w my CBL & db somewhere in b/w. I haven't yet tested it, but I will try to replicate from a cloud. 
Meanwhile, works like a charm with 1.3 .
Do you have a release date yet?

**Comment by snej** [comment]
Interesting … if you can't figure this out, maybe @tleyden knows what's up; he's a Docker guru.

We released the 1.3 dev preview earlier this week and are hoping for a GA in two weeks or so, assuming no showstoppers come up.

**Comment by tleyden** [comment]
@abhishek-shori can you try to run CouchDB directly on OSX and see if you get the same issue?

**Comment by abhishek-shori** [comment]
@snej amazing, Looking forward to it. Thanks!
@tleyden yea, it worked. But, I can't make it work with Docker, Do you have any idea, why should it fail?

**Comment by pasin** [comment]
Is this the same issue as https://github.com/couchbase/couchbase-lite-ios/issues/1403?

**Comment by pasin** [comment]
Closed as fixed in 1.3.
